### PR TITLE
Resources StorageNodeSet and DatabaseNodeSet

### PR DESCRIPTION
**Pull request by kobzonega**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [x] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Storage/Database CR creates one StatefulSet k8s object which operated pods with YDB processes. 

Issue Number: YDBOPS-8645, YDBOPS-8646

## What is the new behavior?

- We propose new CRDs (Custom Resource Definitions) named `StorageNodeSet` and `DatabaseNodeSet`, which define the parameters of StatefulSets for a specific logical group of YDB nodes. The creation and modification of objects occur through the definition in the Storage/Database object, specifying parameters in the `spec.nodeSet` field. Thus controller used in the ydb-operator performs the CRUD operation for *NodeSet CustomResource objects automatically.

- Performing Reconcile for NodeSet objects should include checking the presence of the StatefulSet resource, matching the desired quantity of replicas with the observed quantity. The status of parent objects when using *NodeSet is determined as the composition of readiness statuses StatefulSet and child Pod objects.

- Operations related to interaction with the YDB cluster, such as Storage initialization, Tenant creation, SelfCheck verification, are performed in the Reconcile process for the Storage/Database object. For this purpose, in the parameters of StatefulSet objects created using NodeSet, the `spec.serviceName` (within one parent object)  must be specified identically.

## Other information

Additional refactoring:

- Update Makefile test scenarios to use existing kind cluster

- Use new const variables `GRPCServiceFQDNFormat` and `InterconnectServiceFQDNFormat` to build URL endpoint and domain strings

- Edit const variable `DefaultDatabaseDomain` value from "root" to "Root"

- Remove [database_defaults.go](https://github.com/ydb-platform/ydb-kubernetes-operator/pull/169/files#diff-d37aabd61f17662a9d0d01dffa2c8fcd11e50ab1753f5056a81ea07ad65e8b56) as useless (same logic as in kubebuilder webhook.Defaulter interface)

- Use type `NamespacedRef` for `StorageClusterRef` and `DatabaseRef` fields in CustomResources

- Add new field `StorageEndpoint` to Database and DatabaseNodeSet objects spec to define external FQDN for `--node-broker` container arg

- Remove useless `PublicHost` field in Database CR (use `.spec.Service.GRPC.ExternalHost` and `.spec.Service.Interconnect.ExternalHost` instead)

- Move YDB configuration build logic to a webhook so that it can be generated and specified when ydb-operator mutating API objects

### CODE REVIEW

File: internal/controllers/storage/init.go [hunk 1]
```diff
@@ -109,7 +110,25 @@ func (r *Reconciler) initializeStorage(
...
+	podList := &corev1.PodList{}
+	matchingLabels := client.MatchingLabels{}
+	for k, v := range storage.Labels {
+		matchingLabels[k] = v
+	}
+	opts := []client.ListOption{
+		client.InNamespace(storage.Namespace),
+		matchingLabels,
+	}
+	if err := r.List(ctx, podList, opts...); err != nil || len(podList.Items) == 0 {
```
[the remark this answers is not in the document]
Comment (artgromov) [reply internal/controllers/storage/init.go:123]: Decided to keep that way in this PR.

File: api/v1alpha1/database_webhook.go [hunk 2]
```diff
@@ -145,10 +166,20 @@ func (r *Database) ValidateUpdate(old runtime.Object) error {
 		return errors.New("database domain cannot be changed")
 	}
 
-	if GetDatabasePath(oldDatabase) != GetDatabasePath(r) {
+	if oldDatabase.GetDatabasePath() != r.GetDatabasePath() {
 		return errors.New("database path cannot be changed")
 	}
 
+	if r.Spec.NodeSet != nil {
```
[the remark this answers is not in the document]
Comment (kobzonega) [reply api/v1alpha1/database_webhook.go]: fixed

File: api/v1alpha1/databasenodeset_types.go [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+	StorageDomains []string `json:"storageDomains"`
+
+	// (Optional) Storage services parameter overrides
+	// Default: (not specified)
+	// +optional
+	Service DatabaseServices `json:"service,omitempty"`
+
+	// Encryption
+	// +optional
+	Encryption *EncryptionConfig `json:"encryption,omitempty"`
```
[the remark this answers is not in the document]
Comment (kobzonega) [reply api/v1alpha1/databasenodeset_types.go]: Decided to keep it that way in this PR. Refactor this kind of fields in future

File: api/v1alpha1/databasenodeset_types.go [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+	// +kubebuilder:default:="root"
+	// +optional
+	Domain string `json:"domain"`
+
+	// (Optional) Custom database path in schemeshard
+	// Default: /<spec.domain>/<metadata.name>
+	// +kubebuilder:validation:Pattern:=/[a-zA-Z0-9]([-_a-zA-Z0-9]*[a-zA-Z0-9])?/[a-zA-Z0-9]([-_a-zA-Z0-9]*[a-zA-Z0-9])?(/[a-zA-Z0-9]([-_a-zA-Z0-9]*[a-zA-Z0-9])?)*
+	// +kubebuilder:validation:MaxLength:=255
+	// +optional
+	Path string `json:"path,omitempty"`
```
[the remark this answers is not in the document]
Comment (kobzonega) [reply api/v1alpha1/databasenodeset_types.go]: Decided to keep it that way in this PR. Refactor this kind of fields in future

File: api/v1alpha1/storagenodeset_types.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+	// +optional
+	Configuration string `json:"configuration"`
+
+	// Data storage mode.
+	// For details, see https://cloud.yandex.ru/docs/ydb/oss/public/administration/deploy/production_checklist#topologiya
+	// TODO English docs link
+	// FIXME mirror-3-dc is only supported with external configuration
+	// +kubebuilder:validation:Enum=mirror-3-dc;block-4-2;none
+	// +kubebuilder:default:=block-4-2
+	Erasure ErasureType `json:"erasure"`
```
[the remark this answers is not in the document]
Comment (kobzonega) [reply api/v1alpha1/storagenodeset_types.go]: Decided to keep it that way in this PR. Refactor this kind of fields in future

File: api/v1alpha1/storage_types.go [hunk 6]
```diff
@@ -29,6 +29,10 @@ type StorageSpec struct {
 	// +required
 	DataStore []corev1.PersistentVolumeClaimSpec `json:"dataStore"`
 
+	// (Optional) NodeSet inline configuration to split into multiple StatefulSets
+	// +optional
+	NodeSet []*StorageNodeSetSpecInline `json:"nodeSet,omitempty"`
```
[the remark this answers is not in the document]
Comment (kobzonega) [reply api/v1alpha1/storage_types.go]: Fixed

**Review by artgromov — APPROVED** [review]

**Pull request merged**